homeassistant/components/calendar/__init__.py

```diff
@@ -180,14 +174,37 @@ def update(self):
         else:
             offset_time = dt.dt.timedelta()  # default it
 
+        normalized_event = {}
+
         # cleanup the string so we don't have a bunch of double+ spaces
-        self._cal_data['message'] = re.sub('  +', '', summary).strip()
-        self._cal_data['offset_time'] = offset_time
-        self._cal_data['location'] = self.data.event.get('location', '')
-        self._cal_data['description'] = self.data.event.get('description', '')
-        self._cal_data['start'] = start
-        self._cal_data['end'] = end
-        self._cal_data['all_day'] = 'date' in self.data.event['start']
+        normalized_event['message'] = re.sub('  +', '', summary).strip()
+        normalized_event['location'] = event.get('location', '')
+        normalized_event['description'] = event.get('description', '')
+        normalized_event['start'] = start
+        normalized_event['end'] = end
+        normalized_event['offset_reached'] = self.offset_reached(
+                                                event, offset_time)
+        normalized_event['all_day'] = 'date' in event['start']
+
+        return normalized_event
+
+    def update(self):
+        """Search for the next event."""
+        if not self.data or not self.data.update():
+            # update cached, don't do anything
+            return
+
+        if not self.data.event:
+            # we have no event to work on, make sure we're clean
+            self.cleanup()
+        else:
+            self._cal_data = self.normalize_event(self.data.event)
+
+        if self.data.events:
+            # save all of the events
+            for event in self.data.events:
```

Comment:
balloob suggested using web sockets instead of to get events data

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/system_health/__init__.py#L55-L75

https://developers.home-assistant.io/docs/en/frontend_add_websocket_api.html